### PR TITLE
[SPARK-49748][CORE] Add `getCondition` and deprecate `getErrorClass` in `SparkThrowable`

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
1. Deprecate the `getErrorClass` method in `SparkThrowable`
2. Add new method `getCondition` as the replacement of `getErrorClass` to the `SparkThrowable` interface
3. Use `getCondition` instead of `getErrorClass` in implementations of `SparkThrowable` to avoid warnings.

### Why are the changes needed?
To follow new naming convention proposed by SPARK-46810.

### Does this PR introduce _any_ user-facing change?
Yes, it extends existing APIs.

### How was this patch tested?
By running the existing tests.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by MaxGekk** [comment]
@srielau @panbingkun @nchammas @cloud-fan Could you review the PR, please.

**Comment by nchammas** [comment]
Just noting for reference that my previous attempt to do some of this work is captured in https://github.com/apache/spark/pull/46543. You may find the list of tasks/TODOs there handy.

**Comment by MaxGekk** [comment]
> Just noting for reference that my previous attempt to do some of this work is captured in https://github.com/apache/spark/pull/46543.

@nchammas Sorry, I was on vacation at the time, and wasn't able to review your PR.

> You may find the list of tasks/TODOs there handy.

How about to open sub-tasks in JIRA, and work on the tasks together if you have time. It would be nice to finish the work before the release 4.0.

**Review by srielau — APPROVED** [review]

File: common/utils/src/main/java/org/apache/spark/SparkThrowable.java [hunk 1]
```diff
@@ -35,19 +35,29 @@
...
 @Evolving
 public interface SparkThrowable {
-  // Succinct, human-readable, unique, and consistent representation of the error category
-  // If null, error class is not set
-  String getErrorClass();
+  /**
+   * Succinct, human-readable, unique, and consistent representation of the error condition.
+   * If null, error condition is not set.
+   */
+  String getCondition();
```
Comment (cloud-fan) [inline common/utils/src/main/java/org/apache/spark/SparkThrowable.java:42]: shall we say `getErrorCondition`?

Comment (MaxGekk) [reply common/utils/src/main/java/org/apache/spark/SparkThrowable.java:42]: IMHO, the `error` word is unnecessary. The interface is about errors, and I don't think we should add `error` everywhere. In this logic we should add the error to all methods:
- `getErrorMessageParameters`
- `getErrorQueryContext`
- `getErrorSqlContext`

And when you call `exception.getCondition`, do you expect a condition of somewhat else of the exception/error?

File: common/utils/src/main/java/org/apache/spark/SparkThrowable.java [hunk 2]
```diff
@@ -35,19 +35,29 @@
...
+
+  /**
+   * Succinct, human-readable, unique, and consistent representation of the error category.
+   * If null, error class is not set.
+   * @deprecated Use {@link #getCondition()} instead.
+   */
+  @Deprecated
+  default String getErrorClass() { return getCondition(); }
```
Comment (panbingkun) [inline common/utils/src/main/java/org/apache/spark/SparkThrowable.java:50]: I'm curious if we need to add a rule in `scalastyle-config.xml` that prohibits `getErrorClass` calls? like:

https://github.com/apache/spark/blob/44ec70f5103fc5674497373ac5c23e8145ae5660/scalastyle-config.xml#L145-L151

Comment (MaxGekk) [reply common/utils/src/main/java/org/apache/spark/SparkThrowable.java:50]: I think the warnings about deprecated method is enough for now.

File: common/utils/src/main/java/org/apache/spark/SparkThrowable.java (hunk 1 above)
Comment (MaxGekk) [reply common/utils/src/main/java/org/apache/spark/SparkThrowable.java:42]: @srielau WDYT?

Comment (MaxGekk) [reply common/utils/src/main/java/org/apache/spark/SparkThrowable.java:42]: BTW, the SQL standard just uses the word `condition`:
Screenshot 2024-10-09 at 19 38 51

**Review by cloud-fan — APPROVED** [review]

**Comment by MaxGekk** [comment]
Merging to master. Thank you, @srielau @panbingkun @cloud-fan for review.

**Pull request closed without merging**